tests/unit/api/auth.py

```diff
@@ -0,0 +1,553 @@
+    [(api.node_connect_network, None,
+      False, project,
+      [node, 'boot-nic', net]) for (project, node, net) in [
+          ('runway', 'runway_node_0', 'runway_pxe'),
+          ('runway', 'runway_node_1', 'runway_provider'),
+          ('manhattan', 'manhattan_node_0', 'manhattan_pxe'),
+          ('manhattan', 'manhattan_node_1', 'manhattan_provider'),
+      ]] +
+
+    ### Projects should be able to connect their nodes to public networks.
+    [(api.node_connect_network, None,
+      False, project,
+      [node, 'boot-nic', net]) for (project, node) in [
+          ('runway', 'runway_node_0'),
+          ('runway', 'runway_node_1'),
+          ('manhattan', 'manhattan_node_0'),
+          ('manhattan', 'manhattan_node_1'),
+      ] for net in ('stock_int_pub', 'stock_ext_pub')] +
+
+     ## Illegal cases
+
+     ### Projects should not be able to connect their nodes to each other's
+     ### networks.
+     [(api.node_connect_network, ProjectMismatchError,
+       False, 'runway',
+       [node, 'boot-nic', net]) for (node, net) in [
+           ('runway_node_0', 'manhattan_pxe'),
+           ('runway_node_1', 'manhattan_provider'),
+       ]] +
+
+[
+    ### Projects should not be able to attach each other's nodes to public networks.
+    (api.node_connect_network, AuthorizationError,
+       False, 'runway',
+       ['manhattan_node_0', 'boot-nic', 'stock_int_pub']),
+
+    ### Projects should not be able to attach free nodes to networks.
+    ### The same node about the exception as above applies.
+    (api.node_connect_network, ProjectMismatchError,
+     False, 'runway',
+     ['free_node_0', 'boot-nic', 'stock_int_pub']),
+
+    # list_project_nodes
+
+    ## Legal: admin lists a project's nodes.
+    (api.list_project_nodes, None,
+     True, None,
+     ['runway']),
+
+    ## Legal: project lists its own nodes.
+    (api.list_project_nodes, None,
+     False, 'runway',
+     ['runway']),
+
+    ## Illegal: project lists another project's nodes.
+    (api.list_project_nodes, AuthorizationError,
+     False, 'runway',
+     ['manhattan']),
+
+    # show_node
+
+    ## Legal: project shows a free node
+    (api.show_node, None,
+     False, 'runway',
+     ['free_node_0']),
+
+    ## Legal: project shows its own node.
+    (api.show_node, None,
+     False, 'runway',
+     ['runway_node_0']),
+
+    ## Illegal: project tries to show another project's node.
+    (api.show_node, AuthorizationError,
+     False, 'runway',
+     ['manhattan_node_0']),
+
+    # project_connect_node: Project tries to connect someone else's node
+    # to itself. The basic cases of connecting a free node are covered by
+    # project_calls, below.
+    (api.project_connect_node, BlockedError,
+     False, 'runway',
+     ['runway', 'manhattan_node_0']),
+])
+def test_auth_call(fn, error, admin, project, args):
+    return auth_call_test(fn, error, admin, project, args)
+
```

Comment:
Oh dear lord!  I see what you're doing here, and I see why you're doing it... but we should figure out if this is really the clearest way to do it.  If at least we could have something in that long list besides tuples (dicts, perhaps), that might be better.  I'll think about it for a bit...